p11-kit/test-rpc3.c

```diff
@@ -0,0 +1,771 @@
+	assert_num_eq (0, count);
+
+	count = 8;
+	rv = (rpc_module->C_GetSlotList) (CK_TRUE, slot_list, &count);
+	assert (rv == CKR_OK);
+	assert_num_eq (0, count);
+
+	count = 8;
+	rv = (rpc_module->C_GetSlotList) (CK_FALSE, slot_list, &count);
+	assert (rv == CKR_OK);
+	assert_num_eq (0, count);
+
+	teardown_mock_module (rpc_module);
+}
+
+static void *
+invoke_in_thread (void *arg)
+{
+	CK_FUNCTION_LIST *rpc_module = arg;
+	CK_INFO info;
+	CK_RV rv;
+
+	rv = (rpc_module->C_GetInfo) (&info);
+	assert_num_eq (rv, CKR_OK);
+
+	assert (memcmp (info.manufacturerID, MOCK_INFO.manufacturerID,
+	                sizeof (info.manufacturerID)) == 0);
+
+	return NULL;
+}
+
+static p11_mutex_t delay_mutex;
+
+static CK_RV
+delayed_C_GetInfo (CK_INFO_PTR info)
+{
+	CK_RV rv;
+
+	p11_sleep_ms (rand () % 100);
+
+	p11_mutex_lock (&delay_mutex);
+	rv = mock_C_GetInfo (info);
+	p11_mutex_unlock (&delay_mutex);
+
+	return rv;
+}
+
+static void
+test_simultaneous_functions (void)
+{
+	CK_FUNCTION_LIST_3_0 real_module;
+	CK_FUNCTION_LIST *rpc_module;
+	const int num_threads = 128;
+	p11_thread_t threads[num_threads];
+	int i, ret;
+
+	p11_mutex_init (&delay_mutex);
+
+	memcpy (&real_module, &mock_module_v3_no_slots, sizeof (CK_FUNCTION_LIST));
+	real_module.C_GetInfo = delayed_C_GetInfo;
+
+	rpc_module = setup_test_rpc_module (&test_normal_vtable,
+	                                    &real_module, NULL);
+
+	/* Make the invoked function (above) wait */
+	p11_mutex_lock (&delay_mutex);
+
+	for (i = 0; i < num_threads; i++) {
+		ret = p11_thread_create (threads + i, invoke_in_thread, rpc_module);
+		assert_num_eq (0, ret);
+	}
+
+	/* Let the invoked functions return */
+	p11_mutex_unlock (&delay_mutex);
+
+	for (i = 0; i < num_threads; i++)
+		p11_thread_join (threads[i]);
+
+	teardown_mock_module (rpc_module);
+	p11_mutex_uninit (&delay_mutex);
+}
+
+#ifdef OS_UNIX
+
+static void
+test_fork_and_reinitialize (void)
+{
+	CK_FUNCTION_LIST *rpc_module;
+	CK_INFO info;
+	int status;
+	CK_RV rv;
+	pid_t pid;
+	int i;
+
+	rpc_module = setup_test_rpc_module (&test_normal_vtable,
+	                                    &mock_module_v3_no_slots, NULL);
+
+	pid = fork ();
+	assert_num_cmp (pid, >=, 0);
+
+	/* The child */
+	if (pid == 0) {
+		rv = (rpc_module->C_Initialize) (NULL);
+		assert_num_eq (CKR_OK, rv);
+
+		for (i = 0; i < 32; i++) {
+			rv = (rpc_module->C_GetInfo) (&info);
+			assert_num_eq (CKR_OK, rv);
+		}
+
+		rv = (rpc_module->C_Finalize) (NULL);
+		assert_num_eq (CKR_OK, rv);
+
+		_exit (66);
+	}
+
+	for (i = 0; i < 128; i++) {
+		rv = (rpc_module->C_GetInfo) (&info);
+		assert_num_eq (CKR_OK, rv);
+	}
+
+	assert_num_eq (waitpid (pid, &status, 0), pid);
+	assert_num_eq (WEXITSTATUS (status), 66);
+
+	teardown_mock_module (rpc_module);
+}
+
+#endif /* OS_UNIX */
+
+#include "test-mock.c"
+
+CK_VERSION test_version_three = {CRYPTOKI_VERSION_MAJOR, CRYPTOKI_VERSION_MINOR};
+
+int
+main (int argc,
+      char *argv[])
+{
+	CK_MECHANISM_TYPE mechanisms[] = {
```

Comment:
Let's use `static const` so the array will be part of the text segment instead of a stack. Also I have a little preference on defining it outside the function.

Reply:
Cant be const. But I can move it outside and make it static.